pkg/controller/endpointslicemirroring/reconciler.go

```diff
@@ -0,0 +1,405 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslicemirroring
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	mirroringmetrics "k8s.io/kubernetes/pkg/controller/endpointslicemirroring/metrics"
+	endpointutil "k8s.io/kubernetes/pkg/controller/util/endpoint"
+)
+
+// reconciler is responsible for transforming current EndpointSlice state into
+// desired state
+type reconciler struct {
+	client               clientset.Interface
+	maxEndpointsPerSlice int32
+	endpointSliceTracker *endpointSliceTracker
+	metricsCache         *mirroringmetrics.Cache
+}
+
+// reconcile takes an Endpoints resource and ensures that corresponding
+// EndpointSlices exist. It creates, updates, or deletes EndpointSlices to
+// ensure the desired set of addresses are represented by EndpointSlices.
+func (r *reconciler) reconcile(endpoints *corev1.Endpoints, existingSlices []*discovery.EndpointSlice) error {
+	slicesToCreate := []*discovery.EndpointSlice{}
+	slicesToUpdate := []*discovery.EndpointSlice{}
+	slicesToDelete := []*discovery.EndpointSlice{}
+
+	// Build data structures for existing state.
+	existingSlicesByKey := map[addrTypePortMapKey][]*discovery.EndpointSlice{}
+	numExistingEndpoints := 0
+	for _, existingSlice := range existingSlices {
+		epHash := newAddrTypePortMapKey(existingSlice.Ports, existingSlice.AddressType)
+		existingSlicesByKey[epHash] = append(existingSlicesByKey[epHash], existingSlice)
+		numExistingEndpoints += len(existingSlice.Endpoints)
+	}
+
+	// Build data structures for desired state.
+	desiredPortsByKey := map[addrTypePortMapKey][]discovery.EndpointPort{}
```

Comment:
It feels like the below loop can be made easier to understand (and test) with a small utility object:

```
type desired struct {
  portsByKey map[addrTypePortMapKey][]discovery.EndpointPort{}
  endpointsByKey map[addrTypePortMapKey]endpointSet{}
  numDesiredEndpoints int
}
 
func (d*desired) updatePorts(endpointPorts) { ... }
func (d*desired) addAddr(a) { ... }
func (d*desired) addNotReadyAddr(a) { ... }
```

then the loop becomes:

```
d := newDesired()
for _, subset := range endpoints.Subsets {
  d.updatePorts(epPortsToEpsPorts...)
  for _, a := range subset.Addresses {
    d.addAddress(a)
  }
  for _, a := range subset.NotReadyAddresses {
    d.addNotReadyAddr(a)
  }
}
```

Reply:
Good idea. It didn't turn out quite as elegant thanks to some additional complexity, but I think the end result is still an improvement.